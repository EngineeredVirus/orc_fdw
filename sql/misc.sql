/*-------------------------------------------------------------------------
 *
 * misc.sql
 *    Different types of tests including explain and unsupported features
 *
 *    REQUIRES:
 *      ORC_FDW_DIR variable to be set in the shell which is initiating
 *      the regression.
 *
 * 2020, Hamid Quddus Akhtar.
 *
 * Copyright (c) 2020, Highgo Software Inc.
 *
 * IDENTIFICATION
 *    sql/misc.sql
 *
 *-------------------------------------------------------------------------
 */

\set orc_sample_dir     `echo ${ORC_FDW_DIR}/sample/data`
\set orc_no_format      `echo ${ORC_FDW_DIR}/sample/data/errs`
\set orc_dir_not_valid  `echo ${ORC_FDW_DIR}/sample/data/xyz103`
\set orc_dir_sql        `echo ${ORC_FDW_DIR}/sql`
\set orc_no_such_file   `echo ${ORC_FDW_DIR}/nosuchfile.orc`
\set file_myfile        `echo ${ORC_FDW_DIR}/sample/data/myfile.orc`

/* Create extension */
CREATE EXTENSION orc_fdw;

/* Create server */
CREATE SERVER orc_srv FOREIGN DATA WRAPPER orc_fdw;

/* Let's check version */
select orc_fdw_version();

/* Import all files with .orc extension */
IMPORT
FOREIGN SCHEMA :"orc_sample_dir"
FROM    SERVER orc_srv
INTO    public;

/* Explain */
EXPLAIN VERBOSE
SELECT  *
FROM    myfile;

EXPLAIN VERBOSE
SELECT  x
FROM    myfile;

EXPLAIN VERBOSE
SELECT  y
FROM    myfile;

EXPLAIN VERBOSE
SELECT  x
FROM    myfile
WHERE   y < 10;

EXPLAIN VERBOSE
SELECT  y
FROM    myfile
WHERE   x < 10;

EXPLAIN VERBOSE
SELECT  *
FROM    orc_file_11_format;

EXPLAIN VERBOSE
SELECT  boolean1
        , CASE  boolean1
            WHEN  true    THEN 'yes'
            WHEN  false   THEN 'no'
            ELSE  'never'
          END
        , short1 % 5 AS short1_mod5
        , int1
        , long1
        , float1
        , double1
        , bytes1
        , string1
        , ts
        , decimal1 
FROM    orc_file_11_format
LIMIT   2;

/* Error checking */
IMPORT
FOREIGN SCHEMA :"orc_dir_not_valid"
FROM    SERVER orc_srv
INTO    public;

IMPORT
FOREIGN SCHEMA :"orc_dir_sql"
FROM    SERVER orc_srv
INTO    public;

IMPORT
FOREIGN SCHEMA :"orc_no_format"
FROM    SERVER orc_srv
INTO    public;

CREATE FOREIGN TABLE orc_no_option
(
    x       INT
    , y     INT
)
SERVER orc_srv;

CREATE FOREIGN TABLE orc_no_such_file
(
    x       INT
    , y     INT
)
SERVER orc_srv OPTIONS
(
    FILENAME :'orc_no_such_file'
);

CREATE FOREIGN TABLE myfile_invalid_option
(
    x       INT
    , y     INT
)
SERVER orc_srv OPTIONS
(
    FILENAME :'file_myfile'
    , INVALID_OPTION 'error'
);

/* Unsupported features */
INSERT
INTO    myfile
VALUES  (19, 1039);

UPDATE  myfile
SET     x = -10
WHERE   x = 1;

DELETE
FROM    myfile
WHERE   x = 1;

ANALYZE myfile;

/* Cleanup */
DROP EXTENSION orc_fdw CASCADE;
