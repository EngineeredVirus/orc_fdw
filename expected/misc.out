/*-------------------------------------------------------------------------
 *
 * misc.sql
 *    Different types of tests including explain and unsupported features
 *
 *    REQUIRES:
 *      ORC_FDW_DIR variable to be set in the shell which is initiating
 *      the regression.
 *
 * 2020, Hamid Quddus Akhtar.
 *
 * Copyright (c) 2020, Highgo Software Inc.
 *
 * IDENTIFICATION
 *    sql/misc.sql
 *
 *-------------------------------------------------------------------------
 */
\set orc_sample_dir     `echo ${ORC_FDW_DIR}/sample/data`
\set orc_no_format      `echo ${ORC_FDW_DIR}/sample/data/errs`
\set orc_dir_not_valid  `echo ${ORC_FDW_DIR}/sample/data/xyz103`
\set orc_dir_sql        `echo ${ORC_FDW_DIR}/sql`
\set orc_no_such_file   `echo ${ORC_FDW_DIR}/nosuchfile.orc`
\set file_myfile        `echo ${ORC_FDW_DIR}/sample/data/myfile.orc`
/* Create extension */
CREATE EXTENSION orc_fdw;
/* Create server */
CREATE SERVER orc_srv FOREIGN DATA WRAPPER orc_fdw;
/* Let's check version */
select orc_fdw_version();
    orc_fdw_version     
------------------------
 Highgo ORC FDW - 1.0.0
(1 row)

/* Import all files with .orc extension */
IMPORT
FOREIGN SCHEMA :"orc_sample_dir"
FROM    SERVER orc_srv
INTO    public;
INFO:  ORC file myfile.orc found for schema import.
INFO:  ORC file decimal.orc found for schema import.
INFO:  ORC file orc_file_11_format.orc found for schema import.
WARNING:  orc_fdw: Unsupported ORC file /sources/PG/work/orc_fdw_github/sample/data/orc_file_11_format.orc version 0.11.
HINT:  This may still work, but it's strongly recommended to use files that are supported by the fdw.
INFO:  Schema read successfully from 3 orc files.
/* Explain */
EXPLAIN VERBOSE
SELECT  *
FROM    myfile;
                               QUERY PLAN                                
-------------------------------------------------------------------------
 Foreign Scan on public.myfile  (cost=100.00..600.00 rows=10000 width=8)
   Output: x, y
   ORC File Reader Columns: x, y
(3 rows)

EXPLAIN VERBOSE
SELECT  x
FROM    myfile;
                               QUERY PLAN                                
-------------------------------------------------------------------------
 Foreign Scan on public.myfile  (cost=100.00..600.00 rows=10000 width=4)
   Output: x
   ORC File Reader Columns: x
(3 rows)

EXPLAIN VERBOSE
SELECT  y
FROM    myfile;
                               QUERY PLAN                                
-------------------------------------------------------------------------
 Foreign Scan on public.myfile  (cost=100.00..600.00 rows=10000 width=4)
   Output: y
   ORC File Reader Columns: y
(3 rows)

EXPLAIN VERBOSE
SELECT  x
FROM    myfile
WHERE   y < 10;
                               QUERY PLAN                                
-------------------------------------------------------------------------
 Foreign Scan on public.myfile  (cost=100.00..600.00 rows=10000 width=4)
   Output: x
   Filter: (myfile.y < 10)
   ORC File Reader Columns: x, y
(4 rows)

EXPLAIN VERBOSE
SELECT  y
FROM    myfile
WHERE   x < 10;
                               QUERY PLAN                                
-------------------------------------------------------------------------
 Foreign Scan on public.myfile  (cost=100.00..600.00 rows=10000 width=4)
   Output: y
   Filter: (myfile.x < 10)
   ORC File Reader Columns: x, y
(4 rows)

EXPLAIN VERBOSE
SELECT  *
FROM    orc_file_11_format;
WARNING:  orc_fdw: Unsupported ORC file /sources/PG/work/orc_fdw_github/sample/data/orc_file_11_format.orc version 0.11.
HINT:  This may still work, but it's strongly recommended to use files that are supported by the fdw.
                                                   QUERY PLAN                                                    
-----------------------------------------------------------------------------------------------------------------
 Foreign Scan on public.orc_file_11_format  (cost=100.00..475.00 rows=7500 width=133)
   Output: boolean1, byte1, short1, int1, long1, float1, double1, bytes1, string1, ts, decimal1
   ORC File Reader Columns: boolean1, byte1, short1, int1, long1, float1, double1, bytes1, string1, ts, decimal1
(3 rows)

EXPLAIN VERBOSE
SELECT  boolean1
        , CASE  boolean1
            WHEN  true    THEN 'yes'
            WHEN  false   THEN 'no'
            ELSE  'never'
          END
        , short1 % 5 AS short1_mod5
        , int1
        , long1
        , float1
        , double1
        , bytes1
        , string1
        , ts
        , decimal1 
FROM    orc_file_11_format
LIMIT   2;
WARNING:  orc_fdw: Unsupported ORC file /sources/PG/work/orc_fdw_github/sample/data/orc_file_11_format.orc version 0.11.
HINT:  This may still work, but it's strongly recommended to use files that are supported by the fdw.
                                                                                                             QUERY PLAN                                                                                                              
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=100.00..100.11 rows=2 width=165)
   Output: boolean1, (CASE boolean1 WHEN CASE_TEST_EXPR THEN 'yes'::text WHEN (NOT CASE_TEST_EXPR) THEN 'no'::text ELSE 'never'::text END), (((short1)::integer % 5)), int1, long1, float1, double1, bytes1, string1, ts, decimal1
   ->  Foreign Scan on public.orc_file_11_format  (cost=100.00..512.50 rows=7500 width=165)
         Output: boolean1, CASE boolean1 WHEN CASE_TEST_EXPR THEN 'yes'::text WHEN (NOT CASE_TEST_EXPR) THEN 'no'::text ELSE 'never'::text END, ((short1)::integer % 5), int1, long1, float1, double1, bytes1, string1, ts, decimal1
         ORC File Reader Columns: boolean1, short1, int1, long1, float1, double1, bytes1, string1, ts, decimal1
(5 rows)

/* Error checking */
IMPORT
FOREIGN SCHEMA :"orc_dir_not_valid"
FROM    SERVER orc_srv
INTO    public;
ERROR:  could not open directory "/sources/PG/work/orc_fdw_github/sample/data/xyz103": No such file or directory
IMPORT
FOREIGN SCHEMA :"orc_dir_sql"
FROM    SERVER orc_srv
INTO    public;
INFO:  No files processed.
HINT:  Did you specify the correct folder path with .orc files?
IMPORT
FOREIGN SCHEMA :"orc_no_format"
FROM    SERVER orc_srv
INTO    public;
INFO:  ORC file orc_no_format.orc found for schema import.
WARNING:  orc_fdw: Unsupported ORC file /sources/PG/work/orc_fdw_github/sample/data/errs/orc_no_format.orc version 0.11.
HINT:  This may still work, but it's strongly recommended to use files that are supported by the fdw.
ERROR:  orc_fdw: No supported columns found.
HINT:  Did you specify the correct path for ORC file import? Check documentation to see supported column types.
CREATE FOREIGN TABLE orc_no_option
(
    x       INT
    , y     INT
)
SERVER orc_srv;
ERROR:  orc_fdw: filename option not specified for table.
CREATE FOREIGN TABLE orc_no_such_file
(
    x       INT
    , y     INT
)
SERVER orc_srv OPTIONS
(
    FILENAME :'orc_no_such_file'
);
ERROR:  orc_fdw: filename /sources/PG/work/orc_fdw_github/nosuchfile.orc: No such file or directory
CREATE FOREIGN TABLE myfile_invalid_option
(
    x       INT
    , y     INT
)
SERVER orc_srv OPTIONS
(
    FILENAME :'file_myfile'
    , INVALID_OPTION 'error'
);
ERROR:  orc_fdw: invalid option specified "invalid_option"
/* Unsupported features */
INSERT
INTO    myfile
VALUES  (19, 1039);
ERROR:  orc_fdw: INSERT, UPDATE and DELETE options are not available in this version.
UPDATE  myfile
SET     x = -10
WHERE   x = 1;
ERROR:  orc_fdw: INSERT, UPDATE and DELETE options are not available in this version.
DELETE
FROM    myfile
WHERE   x = 1;
ERROR:  orc_fdw: INSERT, UPDATE and DELETE options are not available in this version.
ANALYZE myfile;
ERROR:  orc_fdw: ANALYZE table options are not available in this version.
/* Cleanup */
DROP EXTENSION orc_fdw CASCADE;
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to server orc_srv
drop cascades to foreign table myfile
drop cascades to foreign table "decimal"
drop cascades to foreign table orc_file_11_format
